### PR TITLE
Fix ghost qubit bug #981

### LINKED ISSUE #981: Ghost qubit in library mode
### Required prerequisites

- [X] Make sure you've read the [documentation](https://nvidia.github.io/cuda-quantum/latest). Your issue may be addressed there.
- [X] Search the [issue tracker](https://github.com/NVIDIA/cuda-quantum/issues) to verify that this hasn't already been reported. +1 or comment there if it has.
- [ ] If possible, make a PR with a failing test to give us a starting point to work on!

### Describe the bug

The sampling of a circuit with a single qubit returns a result with a bitstring with two bits.  

### Steps to reproduce the bug

Compile the following circuit in library mode:
```cpp
#include <cudaq.h>
#include <iostream>

void bar() __qpu__ {
  cudaq::qubit a;
  cudaq::x(a);
  auto a0 = cudaq::mz(a);
  cudaq::reset(a);
  auto a1 = cudaq::mz(a);
}

int main() {
  std::cout << "Bar:\n";
  auto result = cudaq::sample(bar);
  result.dump();

  return 0;
}
```
It's execution returns:
```
Bar:
{ 00:1000 }
```

_Note_: this problem seems to not happen when we add a quantum operator after the `reset`.

### Expected behavior

It should return:
```
Bar:
{ 0:1000 }
```

### Is this a regression? If it is, put the last known working version (or commit) here.

Not a regression

### Environment

- **CUDA Quantum version**: nvq++ proto-0.6.0 (main branch)
- **C++ compiler**: g++ 13.2.0
- **Operating system**:  Ubuntu 23.10 (Docker container)

### Suggestions

_No response_

### PR DESCRIPTION
**Pull request by amccaskey**
Fix #981

### CODE REVIEW

File: unittests/integration/measure_reset_tester.cpp
Comment (bmhowe23) [inline unittests/integration/measure_reset_tester.cpp:25]: Slight optimization since this test requires repeated state vector evolutions and since there is no expected variability in per-shot answers.
```suggestion
  auto result = cudaq::sample(/*shots=*/10, bar);
```

**Review by bmhowe23 — COMMENTED** [review]
I still have a slight reservation about the issue that this PR is attempting to resolve (per [this comment](https://github.com/NVIDIA/cuda-quantum/issues/981#issuecomment-1837182342)), but am ultimately willing to go with whatever the team decides.

That reservation is minor because this change does seem make the behavior of the `__global__` register consistent between library mode and emulation mode, and that aspect is - _no doubt_ - a "good thing" ™️.

File: unittests/integration/measure_reset_tester.cpp
Comment (bmhowe23) [reply unittests/integration/measure_reset_tester.cpp:14]: > When I downloaded this PR to test it locally, this new test is failing on the custatevec-fp32, tensornet, and tensornet-mps backends. Were you able to successfully test it on those backends, @amccaskey? Does this PR have a dependency on another outstanding PR?

To answer my own question, yes, it looks like this PR depends on the change in PR #997 in order to make this new test pass. Once I merged the two of those together, both sets of tests are passing locally.

**Comment by amccaskey** [comment]
> I still have a slight reservation about the issue that this PR is attempting to resolve (per [this comment](https://github.com/NVIDIA/cuda-quantum/issues/981#issuecomment-1837182342)), but am ultimately willing to go with whatever the team decides.
> 
> That reservation is minor because this change does seem make the behavior of the `__global__` register consistent between library mode and emulation mode, and that aspect is - _no doubt_ - a "good thing" ™️.

This PR fixes the bug. We'll figure out as a team how to handle the broader discussion you referenced and that work will need to be part of a separate PR. No reason to hold up this bug fix.

**Review by bmhowe23 — APPROVED** [review]
👍

**Pull request merged**
